File: results/2026-04-17_004319/tasks/13-dependency-license-checker/default-haiku45/generated-code/license_checker.py
"""
Dependency License Checker - Parse manifests and check licenses against allow/deny lists.

TDD Approach:
1. parse_manifest: Extract dependencies from various formats
2. get_license: Mock license lookup (can be replaced with real API)
3. check_licenses: Validate each dependency against config
4. generate_report: Format results for output
"""

import json
import re
from typing import List, Dict, Optional, Any


def parse_manifest(manifest: Any, manifest_type: str = "package.json") -> List[Dict[str, str]]:
    """
    Parse dependency manifest and extract name+version pairs.

    Args:
        manifest: Either dict (for JSON) or string (for text formats)
        manifest_type: Type of manifest ("package.json" or "requirements.txt")

    Returns:
        List of dicts with 'name' and 'version' keys
    """
    if manifest_type == "package.json":
        return _parse_package_json(manifest)
    elif manifest_type == "requirements.txt":
        return _parse_requirements_txt(manifest)
    else:
        raise ValueError(f"Unsupported manifest type: {manifest_type}")


def _parse_package_json(manifest: Dict) -> List[Dict[str, str]]:
    """Extract dependencies from package.json dict."""
    deps = []
    dependencies = manifest.get("dependencies", {})
    for name, version in dependencies.items():
        deps.append({"name": name, "version": version})
    return deps


def _parse_requirements_txt(content: str) -> List[Dict[str, str]]:
    """Extract dependencies from requirements.txt format string."""
    deps = []
    for line in content.strip().split('\n'):
        line = line.strip()
        if not line or line.startswith('#'):
            continue
        # Handle both formats: name==version and name>=version
        # If only name (no operator), use name as-is
        # If operator exists, include operator in version
        match = re.match(r'([a-zA-Z0-9\-_.]+)(==|>=|<=|~=|>|<|!=)?(.*)$', line)
        if match:
            name = match.group(1)
            op = match.group(2) or ""
            version_part = match.group(3) or ""
            # For == operator, just use the version without operator
            # For other operators, include the operator
            if op == "==":
                version = version_part
            else:
                version = f"{op}{version_part}"
            deps.append({"name": name, "version": version})
    return deps


def get_license(name: str, version: str) -> Optional[str]:
    """
    Mock license lookup. In production, this would query an API.
    Returns None if license is unknown.
    """
    # This is a mock - in real scenarios, query npm registry, PyPI, etc.
    return None


def check_licenses(
    dependencies: List[Dict[str, str]],
    config: Dict[str, Any]
) -> List[Dict[str, Any]]:
    """
    Check each dependency's license against allow/deny lists.

    Args:
        dependencies: List of dicts with 'name' and 'version'
        config: Dict with 'allow_licenses' and 'deny_licenses' lists

    Returns:
        List of dicts with license status information
    """
    allow_list = config.get("allow_licenses", [])
    deny_list = config.get("deny_licenses", [])

    report = []
    for dep in dependencies:
        name = dep["name"]
        version = dep["version"]

        license_name = get_license(name, version)

        status_entry = {
            "name": name,
            "version": version,
            "license": license_name
        }

        if license_name is None:
            status_entry["status"] = "unknown"
            status_entry["reason"] = f"Could not determine license for {name}"
        elif license_name in deny_list:
            status_entry["status"] = "denied"
            status_entry["reason"] = f"{license_name} in deny-list"
        elif license_name in allow_list:
            status_entry["status"] = "approved"
        else:
            status_entry["status"] = "unknown"
            status_entry["reason"] = f"{license_name} not in allow-list"

        report.append(status_entry)

    return report


def generate_report(report: List[Dict[str, Any]]) -> str:
    """
    Format compliance report for display.

    Args:
        report: List of dependency status dicts from check_licenses()

    Returns:
        Formatted string report
    """
    lines = [
        "=" * 80,
        "DEPENDENCY LICENSE COMPLIANCE REPORT",
        "=" * 80
    ]

    approved = [d for d in report if d["status"] == "approved"]
    denied = [d for d in report if d["status"] == "denied"]
    unknown = [d for d in report if d["status"] == "unknown"]

    lines.append(f"\nSummary: {len(approved)} approved, {len(denied)} denied, {len(unknown)} unknown")

    if approved:
        lines.append("\n✓ APPROVED:")
        for dep in approved:
            lines.append(f"  {dep['name']}@{dep['version']} [{dep['license']}]")

    if denied:
        lines.append("\n✗ DENIED:")
        for dep in denied:
            lines.append(f"  {dep['name']}@{dep['version']} [{dep['license']}] - {dep.get('reason', '')}")

    if unknown:
        lines.append("\n? UNKNOWN:")
        for dep in unknown:
            lines.append(f"  {dep['name']}@{dep['version']} - {dep.get('reason', '')}")

    lines.append("\n" + "=" * 80)
    return "\n".join(lines)


def load_config(config_path: str) -> Dict[str, Any]:
    """Load license config from JSON file."""
    with open(config_path, 'r') as f:
        return json.load(f)


def main(manifest_path: str, config_path: str, manifest_type: str = "package.json"):
    """
    Main entry point: Load manifest, config, check licenses, output report.
    """
    # Load config
    config = load_config(config_path)

    # Load and parse manifest
    if manifest_type == "package.json":
        with open(manifest_path, 'r') as f:
            manifest = json.load(f)
        dependencies = parse_manifest(manifest, "package.json")
    else:
        with open(manifest_path, 'r') as f:
            content = f.read()
        dependencies = parse_manifest(content, manifest_type)

    # Check licenses
    report = check_licenses(dependencies, config)

    # Generate and print report
    formatted_report = generate_report(report)
    print(formatted_report)

    return report


if __name__ == "__main__":
    import sys
    if len(sys.argv) < 3:
        print("Usage: python license_checker.py <manifest_path> <config_path> [manifest_type]")
        sys.exit(1)

    manifest_path = sys.argv[1]
    config_path = sys.argv[2]
    manifest_type = sys.argv[3] if len(sys.argv) > 3 else "package.json"

    main(manifest_path, config_path, manifest_type)

File: results/2026-04-17_004319/tasks/13-dependency-license-checker/typescript-bun-haiku45/generated-code/license-checker.ts
// Dependency License Checker - Red/Green TDD implementation

// Interfaces for type safety
interface Dependency {
  name: string;
  version: string;
}

interface LicenseConfig {
  allowlist: string[];
  denylist: string[];
}

interface DependencyManifest {
  type: "package.json" | "requirements.txt";
  content: any;
}

type LicenseStatus = "approved" | "denied" | "unknown";

interface DependencyResult {
  name: string;
  version: string;
  license?: string;
  status: LicenseStatus;
}

interface ComplianceReport {
  dependencies: DependencyResult[];
  summary: {
    approved: number;
    denied: number;
    unknown: number;
  };
}

type LicenseLookup = (name: string) => string | undefined;

// Parse dependencies from a manifest
function parseDependencies(manifest: DependencyManifest): Dependency[] {
  if (manifest.type === "package.json") {
    const deps = manifest.content.dependencies || {};
    return Object.entries(deps).map(([name, version]) => ({
      name,
      version: version as string,
    }));
  }
  // Additional manifest types can be added here
  return [];
}

// Check license compliance for dependencies
function checkLicenseCompliance(
  dependencies: Dependency[],
  config: LicenseConfig,
  licenseLookup: LicenseLookup
): ComplianceReport {
  const results: DependencyResult[] = [];
  let approved = 0;
  let denied = 0;
  let unknown = 0;

  for (const dep of dependencies) {
    const license = licenseLookup(dep.name);

    let status: LicenseStatus;
    if (!license) {
      status = "unknown";
      unknown++;
    } else if (config.denylist.includes(license)) {
      status = "denied";
      denied++;
    } else if (config.allowlist.includes(license)) {
      status = "approved";
      approved++;
    } else {
      status = "unknown";
      unknown++;
    }

    results.push({
      name: dep.name,
      version: dep.version,
      license,
      status,
    });
  }

  return {
    dependencies: results,
    summary: { approved, denied, unknown },
  };
}

// Parse requirements.txt format (Python dependencies)
function parseRequirementsTxt(content: string): Dependency[] {
  return content
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line && !line.startsWith("#"))
    .map((line) => {
      // Handle various version specifiers: ==, >=, <=, ~=, etc.
      const match = line.match(/^([a-zA-Z0-9_-]+)(.*)/);
      if (!match) return null;

      const name = match[1];
      // Extract version from any operator (==, >=, <=, ~=, etc.)
      const versionMatch = match[2].match(/^[><=~!]+(.*)$/);
      const version = versionMatch ? versionMatch[1].trim() : "";

      return { name, version };
    })
    .filter((dep): dep is Dependency => dep !== null);
}

// Generate a text report from compliance results
function generateComplianceReport(report: ComplianceReport): string {
  const lines: string[] = [];
  lines.push("=== Dependency License Compliance Report ===\n");

  for (const dep of report.dependencies) {
    const status = dep.status.toUpperCase();
    const license = dep.license || "UNKNOWN";
    lines.push(`[${status}] ${dep.name}@${dep.version} (${license})`);
  }

  lines.push("");
  lines.push(
    `Summary: ${report.summary.approved} approved, ${report.summary.denied} denied, ${report.summary.unknown} unknown`
  );

  return lines.join("\n");
}

export {
  parseDependencies,
  checkLicenseCompliance,
  parseRequirementsTxt,
  generateComplianceReport,
  DependencyManifest,
  Dependency,
  LicenseConfig,
  DependencyResult,
  ComplianceReport,
  LicenseStatus,
  LicenseLookup,
};
